cpu/atmega_common/periph/rtt.c

```diff
@@ -0,0 +1,261 @@
+    TCNT2 = 0;
+
+    /* Reset compare values */
+    OCR2A = 0;
+    OCR2B = 0;
+
+    /* Reset timer control */
+    TCCR2A = 0;
+
+    /* 32768Hz / 1024 = 32 ticks per second */
+    TCCR2B = (1 << CS22) | (1 << CS21) | (1 <<  CS20);
+
+    /* Wait until not busy anymore */
+    DEBUG("RTT waits until ASSR not busy\n");
+    _asynch_wait();
+
+    /* Clear interrupt flags */
+    /* Oddly, this is done by writing ones; see datasheet */
+    TIFR2 = (1 << OCF2B) | (1 << OCF2A) | (1 << TOV2);
+
+    /* Enable 8-bit overflow interrupt */
+    TIMSK2 |= (1 << TOIE2);
+
+    DEBUG("RTT initialized\n");
+}
+
+void rtt_set_overflow_cb(rtt_cb_t cb, void *arg)
+{
+    /* Interrupt safe order of assignment */
+    rtt_state.overflow_cb = NULL;
+    rtt_state.overflow_arg = arg;
+    rtt_state.overflow_cb = cb;
+}
+
+void rtt_clear_overflow_cb(void)
+{
+    /* Interrupt safe order of assignment */
+    rtt_state.overflow_cb = NULL;
+    rtt_state.overflow_arg = NULL;
+}
+
+uint32_t rtt_get_counter(void)
+{
+    /* Make sure it is safe to read TCNT2, in case we just woke up */
+    DEBUG("RTT sleeps until safe to read TCNT2\n");
+    TCCR2A = 0;
+    _asynch_wait();
+
+    return (((uint32_t)rtt_state.ext_cnt << 8) | (uint32_t)TCNT2);
+}
+
+void rtt_set_counter(uint32_t counter)
+{
+    /* Wait until not busy anymore (should be immediate) */
+    DEBUG("RTT sleeps until safe to write TCNT2\n");
+    _asynch_wait();
+
+    rtt_state.ext_cnt = (uint16_t)(counter >> 8);
+    TCNT2 = (uint8_t)counter;
+}
+
+void rtt_set_alarm(uint32_t alarm, rtt_cb_t cb, void *arg)
+{
+    /* Disable alarm interrupt */
+    TIMSK2 &= ~(1 << OCIE2A);
+    rtt_state.alarm_cb = NULL;
+
+    /* Wait until not busy anymore (should be immediate) */
+    DEBUG("RTT sleeps until safe to write OCR2A\n");
+    _asynch_wait();
+
+    /* Set the alarm value */
+    rtt_state.ext_comp = (uint16_t)(alarm >> 8);
+    OCR2A = (uint8_t)alarm;
+
+    /* Interrupt safe order of assignment */
+    rtt_state.alarm_arg = arg;
+    rtt_state.alarm_cb = cb;
+
+    /* Enable alarm interrupt only if it will trigger before overflow */
+    if (rtt_state.ext_comp <= rtt_state.ext_cnt) {
+        TIMSK2 |= (1 << OCIE2A);
+    }
+}
+
+uint32_t rtt_get_alarm(void)
+{
+    return (((uint32_t)rtt_state.ext_comp << 8) | (uint32_t)OCR2A);
+}
+
+void rtt_clear_alarm(void)
+{
+    /* Disable alarm interrupt */
+    TIMSK2 &= ~(1 << OCIE2A);
+
+    /* Interrupt safe order of assignment */
+    rtt_state.alarm_cb = NULL;
+    rtt_state.alarm_arg = NULL;
+}
+
+void rtt_poweron(void)
+{
+    power_timer2_enable();
+}
+
+void rtt_poweroff(void)
+{
+    power_timer2_disable();
+}
+
+void _asynch_wait(void)
+{
+    /* Wait until all busy flags clear. According to the datasheet,
+     * this can take up to 2 positive edges of TOSC1 (32kHz). */
+    while (ASSR & ((1 << TCN2UB) | (1 << OCR2AUB) | (1 << OCR2BUB)
+                   | (1 << TCR2AUB) | (1 << TCR2BUB))) ;
+}
+
+ISR(TIMER2_OVF_vect) {
```

Comment:
Could be reduced futher more as the 16 bit should automatically overflow. 
```
ISR(TIMER2_OVF_vect) {
    __enter_isr();

    /* Enable RTT alarm if overflowed enough times */
    if (++rtt_state.ext_cnt == rtt_state.ext_comp ) {
        TIMSK2 |= (1 << OCIE2A);
    }

#if MODULE_PERIPH_RTC
    /* Increment RTC by 8 seconds */
    atmega_rtc_incr();
#endif

    /* Virtual 24-bit timer overflow */
    if (rtt_state.ext_cnt == 0) {

        /* Execute callback */
        if (rtt_state.overflow_cb != NULL) {
            rtt_state.overflow_cb(rtt_state.overflow_arg);
        }
    }

    if (sched_context_switch_request) {
        thread_yield();
    }
    __exit_isr();
}
```

Reply:
"Should", yes, and would in almost all cases. However, the standard states that the result of an overflow is undefined.

Reply:
This only accounts for signed values. unsigned values are well defined in the C standard.

```
§6.2.5/9
A computation involving unsigned operands can never overflow,
because a result that cannot be represented by the resulting unsigned integer type is
reduced modulo the number that is one greater than the largest value that can be
represented by the resulting type.
```

Reply:
My mistake. Removed the extra logic.